experimental/waveshare_eink/linux_kernel_modules/fb_waveshare_eink.c

```diff
@@ -419,7 +420,7 @@ enum waveshare_devices {
 
 static struct waveshare_eink_device_properties devices[] =
 {
-	[DEV_WS_213] = {.width = 128, .height = 250, .bpp = 1},
+	[DEV_WS_213] = {.width = 122, .height = 250, .bpp = 1},
```

Comment:
I think the easiest solution is to change the struct to have:
u16 xres; // 122 in this case
u16 yres; // 250 in this case
u16 xres_virtual; // 128 in this case
u16 yres_virtual; // 250 in this case
u8 bpp; // 1 in this case

Then adjust the driver to use xres or xres_virtual as appropriate.

Reply:
Hi David.

From user space we using xres to draw. xres should multiple of 8.
So if we using xres=122 display is incorrect.

I think 122 or 128 just related to real resolution and logical resolution
"
// Display resolution
/* the resolution is 122x250 in fact */
/* however, the logical resolution is 128x250 */
#define EPD_WIDTH       128
#define EPD_HEIGHT      250

Reply:
xres yres be visible range size, corresponding to the LCD display size.
xres_virtual yres_virtual defines a framebuffer memory size.

xres_virtual yres_virtual must be greater than or equal to xres yres, pan operation can be displayed and yres_virtual xres_virtual defined display area.

And xres_virtual and y res virtual support for pan in case we using NUMBER_OF_BUFFERS>1
```fb->fb.var.xres		= width;
fb->fb.var.yres		= height;
fb->fb.var.xres_virtual	= width;
fb->fb.var.yres_virtual	= height * ANDROID_NUMBER_OF_BUFFERS;```